### DISCUSSION TITLE
User ID and User Type not being added in audits table

### DISCUSSION THREAD

**Discussion by nabeel-qavi** [owen-it/laravel-auditing#864] [Q&A] [answered]
| Q                 | A
| ----------------- | ---
| Bug?              | no|yes
| New Feature?      | no|yes
| Framework         | Laravel|Lumen
| Framework version | 8.x.y
| Package version   | 13.x.y
| PHP version       | 7.3.y

#### Actual Behaviour

I am working with Laravel Audit Package and i am using laravel 8 version , the problem is that the project is not using any sort of middleware for the api ,It is working but does not enter user_id and user_type columns in audits table, i have tried implementing the scenerios from the documentation but it is not working

#### Expected Behaviour
It should also add user_id and user_type in audits column while is is adding for other columns 

#### Steps to Reproduce

List **all the steps** needed to reproduce the issue you're having.

Make sure to **include code** (affected models, configuration, ...).

** <?php

return [

    'enabled' => env('AUDITING_ENABLED', true),

    /*
    |--------------------------------------------------------------------------
    | Audit Implementation
    |--------------------------------------------------------------------------
    |
    | Define which Audit model implementation should be used.
    |
    */

    'implementation' => OwenIt\Auditing\Models\Audit::class,

    /*
    |--------------------------------------------------------------------------
    | User Morph prefix & Guards
    |--------------------------------------------------------------------------
    |
    | Define the morph prefix and authentication guards for the User resolver.
    |
    */

    'user'      => [
        'morph_prefix' => 'user',
        'guards'       => [
            'web',
            'api'
        ],
        'resolver'     => OwenIt\Auditing\Resolvers\UserResolver::class
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Resolvers
    |--------------------------------------------------------------------------
    |
    | Define the IP Address, User Agent and URL resolver implementations.
    |
    */
    'resolvers' => [
        'ip_address' => OwenIt\Auditing\Resolvers\IpAddressResolver::class,
        'user_agent' => OwenIt\Auditing\Resolvers\UserAgentResolver::class,
        'url'        => OwenIt\Auditing\Resolvers\UrlResolver::class,
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Events
    |--------------------------------------------------------------------------
    |
    | The Eloquent events that trigger an Audit.
    |
    */

    'events' => [
        'created',
        'updated',
        'deleted',
        'restored'
    ],

    /*
    |--------------------------------------------------------------------------
    | Strict Mode
    |--------------------------------------------------------------------------
    |
    | Enable the strict mode when auditing?
    |
    */

    'strict' => false,

    /*
    |--------------------------------------------------------------------------
    | Global exclude
    |--------------------------------------------------------------------------
    |
    | Have something you always want to exclude by default? - add it here.
    | Note that this is overwritten (not merged) with local exclude
    |
    */

    'exclude' => [],

    /*
    |--------------------------------------------------------------------------
    | Empty Values
    |--------------------------------------------------------------------------
    |
    | Should Audit records be stored when the recorded old_values & new_values
    | are both empty?
    |
    | Some events may be empty on purpose. Use allowed_empty_values to exclude
    | those from the empty values check. For example when auditing
    | model retrieved events which will never have new and old values.
    |
    |
    */

    'empty_values'         => true,
    'allowed_empty_values' => [
        'retrieved'
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Timestamps
    |--------------------------------------------------------------------------
    |
    | Should the created_at, updated_at and deleted_at timestamps be audited?
    |
    */

    'timestamps' => false,

    /*
    |--------------------------------------------------------------------------
    | Audit Threshold
    |--------------------------------------------------------------------------
    |
    | Specify a threshold for the amount of Audit records a model can have.
    | Zero means no limit.
    |
    */

    'threshold' => 0,

    /*
    |--------------------------------------------------------------------------
    | Audit Driver
    |--------------------------------------------------------------------------
    |
    | The default audit driver used to keep track of changes.
    |
    */

    'driver' => 'database',

    /*
    |--------------------------------------------------------------------------
    | Audit Driver Configurations
    |--------------------------------------------------------------------------
    |
    | Available audit drivers and respective configurations.
    |
    */

    'drivers' => [
        'database' => [
            'table'      => 'audits',
            'connection' => null,
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Console
    |--------------------------------------------------------------------------
    |
    | Whether console events should be audited (eg. php artisan db:seed).
    |
    */

    'console' => false,
];
** 
The above code is for audit.php

**
<?php

namespace App\AuditResolvers;

use OwenIt\Auditing\Contracts\Auditable;
use OwenIt\Auditing\Contracts\Resolver;

class UserResolver implements Resolver
{
    // public static function resolve(Auditable $auditable)
    // {
    //     //return Auth::check() ? Auth::id() : 2;
    //     return 22;
    // }

    public static function resolveId()
    {
        // Implement your logic to retrieve the authenticated user's ID
        // Here, we'll use Laravel's Auth facade to get the ID
        return Auth::id();
    }

}
**
The above code is for UserResolver which i created
![Screenshot (1)](https://github.com/owen-it/laravel-auditing/assets/143005519/7f709691-6cc0-4292-a8a1-bcd300d50e55)
![Screenshot (2)](https://github.com/owen-it/laravel-auditing/assets/143005519/fdd430be-0e7e-4ec7-99c7-6f8b53abeb18)

#### Possible Solutions

If you have any ideas on how to solve the issue, add them here, otherwise you can omit this part.

**Comment by parallels999** [comment, marked as answer]
User must be logged, if not, you must create your own CustomUserResolver
https://github.com/owen-it/laravel-auditing/blob/7d929c01cad48be1785ae8c7160761d390e2f46b/src/Auditable.php#L333
https://github.com/owen-it/laravel-auditing/blob/7d929c01cad48be1785ae8c7160761d390e2f46b/src/Auditable.php#L341-L342
https://github.com/owen-it/laravel-auditing/blob/master/src/Resolvers/UserResolver.php
